Add a fast abort path for StringImpl's equalInternal()

cdumez opened the pull request:
#### c5bb7e500378f892814484eca2648286358c10ae
<pre>
Add a fast abort path for StringImpl&apos;s equalInternal()
<a href="https://bugs.webkit.org/show_bug.cgi?id=254766">https://bugs.webkit.org/show_bug.cgi?id=254766</a>

Reviewed by Darin Adler and Geoffrey Garen.

Abort early if the first character is not equal instead of calling
equal() with all the characters unconditionally.

* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::equalInternal):

Canonical link: <a href="https://commits.webkit.org/262465@main">https://commits.webkit.org/262465@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed2700dd4a362129bb5a775cb305c2493f1958cd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1405 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1434 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1484 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2301 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1245 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1495 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1494 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1377 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1418 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1299 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1277 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2147 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1263 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1230 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1217 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1278 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1299 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2369 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1363 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1337 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1195 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1442 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1252 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1270 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/298 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/396 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1373 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1446 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/294 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -1417,8 +1417,8 @@ template<typename CharacterType> inline bool equalInternal(const StringImpl* a,
     if (a->is8Bit())
-        return equal(a->characters8(), b, length);
-    return equal(a->characters16(), b, length);
+        return *a->characters8() == *b && equal(a->characters8(), b, length);
```
Maybe the equal call should be

`equal(a->characters8() + 1, b + 1, length -1)`

?

Probably not a huge difference, but might as well?

cdumez replied:
Done.

geoffreygaren reviewed (approved):
r=me

darinadler commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -1416,9 +1416,11 @@ template<typename CharacterType> inline bool equalInternal(const StringImpl* a,
         return false;
+    if (!length)
+        return true;
     if (a->is8Bit())
-        return equal(a->characters8(), b, length);
-    return equal(a->characters16(), b, length);
+        return *a->characters8() == *b && equal(a->characters8() + 1, b + 1, length - 1);
+    return *a->characters16() == *b && equal(a->characters16() + 1, b + 1, length - 1);
```
How did you decide to put the optimization here rather than in the inline `equal(a, b, length)` functions?

cdumez replied:
This is the place where we were doing the is8Bit() check already and I need access to the first character. Given this, it seems like a good place to add the check.

darinadler replied:
Seems to me like this would be even better inside the highly-optimized `ALWAYS_INLINE` `equal` functions, not here just before calling them.

cdumez replied:
The function that is really hot is:
```
WTF_EXPORT_PRIVATE bool equal(const StringImpl*, const LChar*, unsigned); // In StringImpl.h
```

It is not inline. It is defined in the cpp like so:
```
bool equal(const StringImpl* a, const LChar* b, unsigned length)
{
    return equalInternal(a, b, length);
}
```

This is why I modified `equalInternal()`.

```
Sample Count, Samples %, Normalized CPU %, Symbol
483, 0.2%, 0.1%, WTF::equal(WTF::StringImpl const*, unsigned char const*, unsigned int) (in JavaScriptCore)
217, 0.1%, 0.0%,     WTF::AtomStringImpl::add(unsigned char const*, unsigned int) (in JavaScriptCore)
146, 0.1%, 0.0%,     JSC::jsonProtoFuncParse(JSC::JSGlobalObject*, JSC::CallFrame*) (in JavaScriptCore)
33, 0.0%, 0.0%,     WebCore::HTMLFastPathParser<unsigned char>::parseAttributes(WebCore::Element&) (in WebCore)
```

cdumez replied:
Is your suggestion that we try to make part of this function inline and do the quick checks in the inline portion?

darinadler replied:
No.

`StringCommon.h` has `ALWAYS_INLINE` functions `equal(const LChar*, const LChar*, unsigned)` and `equal(const UChar*, const UChar*, unsigned)`. I am suggesting that we do the quick checks in those functions instead of in `equalInternal`.

I don’t think you need to do that in this patch. I am saying that I believe putting the optimizations there _might_ give us even better results, because other call sites likely would get similar benefit. But it’s also possible it might not work as well.

Separately, maybe it is worth exploring a more-inlined version, what you thought I might mean. The fact that we see this function call at all on the profile, given how simple the algorithm is, I think often does mean there is a possible improvement.

cdumez replied:
I think there is a misunderstanding. If you look at the trace (pasted above) and the corresponding code, you'll see that the AtomStringTable code calls `equal(const StringImpl* a, const LChar* b, unsigned length)` directly for example.

Looks to me that moving the checks to the function you suggest may help for other workloads but won't impact the hot ones on Speedometer?

cdumez replied:
`equal(const LChar*, const LChar*, unsigned)` and `equal(const UChar*, const UChar*, unsigned)` are other variants, not using the code path I am modifying in this patch. They are not the ones I see as super high on Speedometer.

I can investigate those separately but unless I am missing something they are not using the code I am modifying and not related to the Speedometer samples I am working on reducing.

darinadler replied:
> Looks to me that moving the checks to the function you suggest may help for other workloads but won't impact the hot ones on Speedometer?

Yes. Sorry if I accidentally seemed to imply otherwise.

No further impact on Speedometer might still mean improvement in other cases. Or moving it might even ruin the Speedometer improvement. But it seems likely to me that most string comparisons could benefit from a fast out for first character mismatches.

I’m **glad** you are fully focused on the actual measurable Speedometer impact, but I was thinking slightly more broadly when making my suggestion.

darinadler replied:
> `equal(const LChar*, const LChar*, unsigned)` and `equal(const UChar*, const UChar*, unsigned)` are other variants, not using the code path I am modifying in this patch. They are not the ones I see as super high on Speedometer.

Wait, what? I am pretty sure that they **are** the functions called in the last two lines of `equalInternal`.

cdumez replied:
OHHHH, indeed :) I got confused, I thought you were talking about one of the callers but those are actually callees.
I see your point better now. I'll do some more A/B testing.

cdumez replied:
And now I realize that if your approach works out, we'd need a single patch instead of at least two:
https://github.com/WebKit/WebKit/pull/12239

cdumez replied:
Doing the check in the ALWAYS_INLINE equal() didn't measure as a progression so I'll land this version for now.

cdumez commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -1416,9 +1416,11 @@ template<typename CharacterType> inline bool equalInternal(const StringImpl* a,
 
     if (a->length() != length)
         return false;
+    if (!length)
```
I had to add this check to address the test failures so I am a/b testing again.

cdumez committed:
Add a fast abort path for StringImpl's equalInternal()
https://bugs.webkit.org/show_bug.cgi?id=254766

Reviewed by Darin Adler and Geoffrey Garen.

Abort early if the first character is not equal instead of calling
equal() with all the characters unconditionally.

* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::equalInternal):

Canonical link: https://commits.webkit.org/262465@main
